### PR TITLE
[FLINK-24227][connectors/kinesis] Moved AWS constants and general utils to common AWS connector

### PR DESCRIPTION
**Pull request by nuno-c-afonso**
## What is the purpose of the change

- The `flink-connector-kinesis` connector contains AWS utilities that could be re-used across other connectors that interact with other AWS services
- This changes moves such utilities to a common connector (`flink-connector-aws-base`) that can be used directly by other connectors

## Brief change log

- Added `flink-connector-aws-base`, which is a new module that contains AWS utilities extracted from `flink-connector-kinesis`

## Verifying this change

This change added tests and can be verified as follows:

- Tests for AWS utilities that were part of `flink-connector-kinesis` are replicated into `flink-connector-aws-base`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by dannycranmer** [comment]
As discussed @nuno-c-afonso please rename the package from `common` to `base` for consistency with other modules

File: flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+
+/** Configuration keys for AWS service usage. */
+@PublicEvolving
+public class AWSConfigConstants {
+
+    /**
+     * Possible configuration values for the type of credential provider to use when accessing AWS
+     * Kinesis. Internally, a corresponding implementation of {@link AwsCredentialsProvider} will be
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java:30]: Can we make this comment AWS generic?

File: flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        WEB_IDENTITY_TOKEN,
+
+        /**
+         * A credentials provider chain will be used that searches for credentials in this order:
+         * ENV_VARS, SYS_PROPS, WEB_IDENTITY_TOKEN, PROFILE in the AWS instance metadata. *
+         */
+        AUTO,
+    }
+
+    /** The AWS region of the Kinesis streams to be pulled ("us-east-1" is used if not set). */
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java:75]: Can we make this comment AWS generic?

File: flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    /**
+     * The credentials provider that provides credentials for assuming the role when credential
+     * provider type is set to ASSUME_ROLE. Roles can be nested, so AWS_ROLE_CREDENTIALS_PROVIDER
+     * can again be set to "ASSUME_ROLE"
+     */
+    public static final String AWS_ROLE_CREDENTIALS_PROVIDER =
+            roleCredentialsProvider(AWS_CREDENTIALS_PROVIDER);
+
+    /** The AWS endpoint for Kinesis (derived from the AWS region setting if not set). */
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java:126]: Can we make this comment AWS generic?

File: flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    public static final String AWS_ENDPOINT = "aws.endpoint";
+
+    /** Whether to trust all SSL certificates. */
+    public static final String TRUST_ALL_CERTIFICATES = "aws.trust.all.certificates";
+
+    /** The HTTP protocol version to use. */
+    public static final String HTTP_PROTOCOL_VERSION = "aws.http.protocol.version";
+
+    /** The identifier of the legacy connector. */
+    public static final String LEGACY_CONNECTOR = "aws.kinesis.legacy";
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/config/AWSConfigConstants.java:136]: Can we make this field AWS generic?

**Review by dannycranmer — CHANGES_REQUESTED** [review]

**Comment by CrynetLogistics** [comment]
e52bcdb Azure: SUCCESS

File: flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+import java.util.Optional;
+import java.util.Properties;
+
+/** Some general utilities specific to Amazon Web Service. */
+@Internal
+public class AWSGeneralUtil {
+    private static final int INITIAL_WINDOW_SIZE_BYTES = 512 * 1024; // 512 KB
+    private static final Duration HEALTH_CHECK_PING_PERIOD = Duration.ofSeconds(60);
+
+    @VisibleForTesting
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/main/java/org/apache/flink/connector/aws/util/AWSGeneralUtil.java:55]: Why have these been made visible for testing? Generally speaking `@VisibleForTesting` is a code smell, and using constants for test assertion is also not good. If someone accidentally modifies this value, the test will still pass.

File: flink-connectors/flink-connector-aws-base/src/test/java/org/apache/flink/connector/aws/util/AWSGeneralUtilTest.java [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+                nettyDefaultConfiguration.tlsTrustManagersProvider(),
+                nettyConfiguration.tlsTrustManagersProvider());
+        assertEquals(
+                AWSGeneralUtil.TRUST_ALL_CERTIFICATES, nettyConfiguration.trustAllCertificates());
+        assertEquals(
+                nettyDefaultConfiguration.writeTimeoutMillis(),
+                nettyConfiguration.writeTimeoutMillis());
+        assertEquals(
+                AWSGeneralUtil.HTTP_PROTOCOL,
+                nettyConfiguration.attribute(SdkHttpConfigurationOption.PROTOCOL));
```
Comment (dannycranmer) [inline flink-connectors/flink-connector-aws-base/src/test/java/org/apache/flink/connector/aws/util/AWSGeneralUtilTest.java:293]: nit: These would be best as single tests. Best practise is to make tests test a single thing, and ideally have a single assertion. This way we fail fast, and get all failures immediately.

**Comment by dannycranmer** [comment]
LGTM, will merge once CI passes

**Comment by dannycranmer** [comment]
CI passed https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=27413&view=results

**Comment by dannycranmer** [comment]
Merged to master https://github.com/apache/flink/commit/c907bae7f4837ca23299e60aca9c8847454b245f

**Pull request closed without merging**